### PR TITLE
Human readable page title

### LINKED ISSUE #1830: Generate document.title based on location pathname
**Is your feature request related to a problem? Please describe.**

At the moment, each page of the Merchant Center shows the `Merchant Center` title. This is not very good for a number of reasons:

- With many tabs open, it's hard to find the correct one, as there is no page context information available.
- Using the browser history navigation is pretty much useless:

image

**Describe the solution you'd like**

Each page should reflect its context in the `document.title` so that users can understand where they are. For that we can derive the information from the URL, as the default behavior. Users can still decide to provide a more precise value by passing a custom render function.

### PR DESCRIPTION
**Pull request by ddouglasz**
Closes #1830 

Page titles should display a human readable title with respect to the user navigation.

### CODE REVIEW

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+const ApplicationPageTitle = <Query extends {} = {}>(props: Props<Query>) => {
+  const location = useLocation();
+  const breadcrumb = compilePageTitle(location as TEnhancedLocation<Query>);
+  const pageTitle = props.renderPageTitle
+    ? props.renderPageTitle(breadcrumb)
+    : `${convertToReadablePrefix(breadcrumb.paths[0])} - ${breadcrumb.suffix}`;
+
+  useEffect(() => {
+    document.title = pageTitle;
```
Comment (ddouglasz) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:124]: I am still going to look to see if there is an alternative to accessing the DOM directly, since that is not so advisable. 
We agreed to not use the helmet package as it seems not to be maintained for a while now.

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    case 'settings':
+      return 'Settings';
+    case 'projects':
+      return 'Projects';
+    case 'login':
+      return 'Login';
+    case 'account':
+      return 'Account';
+    default:
+      return startCase(prefix);
```
Comment (tdeekens) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:112]: 👋🏼 Thanks for working on this. I am frankly new to the topic but am wondering if this is something we should allow configuration of via a Custom Application's config. Then we might not need this map here.

Comment (ddouglasz) [reply packages/application-shell/src/components/application-page-title/application-page-title.tsx:112]: Oh, thanks for pointing this out. I will look into it.

Comment (emmenko) [reply packages/application-shell/src/components/application-page-title/application-page-title.tsx:112]: No need to configure this, this mapping is not really necessary.

I would suggest to focus on this comment first: https://github.com/commercetools/merchant-center-application-kit/pull/2758#issuecomment-1216602555

Then things will fall nicely into place, following that path.

**Comment by ddouglasz** [comment]
> @ddouglasz How can we ensure we're implementing this according to the requirements? 🤓

Ah, great question. As for the default mapping, I think I pretty much tested it to see that it matches and is consistent with all the possible requirements on  what we have in the documented requirements. From what is in the doc, we are either using the `page identifier`, `project key` and `Merchant Center`. 
The paths pattern may change sometimes so accessing some attributes of the paths might change for different locations eg `account`, and `login` as they do not seem to begin with a `project_key` for example. These are some of the cases I tried to cover too. Here default mapping uses  just `login` and `Merchant Center`, or `account` and `Merchant Center`.

**Comment by emmenko** [comment]
> I think I pretty much tested it to see that it matches and is consistent with all the possible requirements on what we have in the documented requirements.

That's good yes, I just don't see that in the PR 😉

File: packages/application-shell/src/components/application-page-title/application-page-title.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  it('should render a truncated readable page title', async () => {
+    render(
+      <Router history={memoryHistory(route)}>
+        <ApplicationPageTitle />
+      </Router>
+    );
+    await waitFor(() => {
+      expect(document.title).toBe(title);
+    });
+  });
```
Comment (CarlosCortizasCT) [inline packages/application-shell/src/components/application-page-title/application-page-title.spec.tsx:63]: nit: perhaps we can extract this logic to a private helper function which can be used in all three tests ([here is an idea](https://gist.github.com/CarlosCortizasCT/ea4aab5614225e6becb19d45c1bb61b6))

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+};
+
+const ApplicationPageTitle = <Query extends {} = {}>(props: Props) => {
+  const location = useLocation();
+  const breadcrumb = compilePageTitle(location as TEnhancedLocation<Query>);
+  const defaultMapping = `${startCase(breadcrumb.paths[0])} - ${
+    breadcrumb.suffix
+  }`;
+  const pageTitle = props.renderPageTitle
+    ? `${props.renderPageTitle
```
Comment (CarlosCortizasCT) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:100]: i guess we would need to also test for an empty array, otherwise the page title will start with a slash:
```suggestion
    ? `${props.renderPageTitle && props.renderPageTitle.length > 0
```

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    : defaultMapping;
+
+  const truncatePageTitle = (pageTitle: string) => {
+    if (pageTitle.length > maxTitleCharLength) {
+      return `${pageTitle.slice(0, 12)} ... ${pageTitle.slice(
+        pageTitle.length - 12
+      )}`;
+    }
+    return pageTitle;
+  };
```
Comment (CarlosCortizasCT) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:112]: This seems to be a pure function so we can move it outside the component so we don't create it on every render.

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+};
+
+const ApplicationPageTitle = <Query extends {} = {}>(props: Props) => {
+  const location = useLocation();
+  const breadcrumb = compilePageTitle(location as TEnhancedLocation<Query>);
+  const defaultMapping = `${startCase(breadcrumb.paths[0])} - ${
+    breadcrumb.suffix
+  }`;
+  const pageTitle = props.renderPageTitle
```
Comment (CarlosCortizasCT) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:99]: I'm wondering if we should move this code block inside the effect as it only needs to run when location or renderPageTitle changes:
```
  useLayoutEffect(() => {
    const breadcrumb = compilePageTitle(location as TEnhancedLocation<Query>);
    const defaultMapping = `${startCase(breadcrumb.paths[0])} - ${
      breadcrumb.suffix
    }`;
    const pageTitle = props.renderPageTitle && props.renderPageTitle.length > 0
      ? `${props.renderPageTitle
          .join(' - ')
          .replace(/\w+/g, capitalize)} - ${defaultMapping}`
      : defaultMapping;

    document.title = truncatePageTitle(pageTitle);
  }, [location, props.renderPageTitle]);
```

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+  const paths = isStaticPath(projectKeyOrStaticPath)
+    ? getLimitedPaths([projectKeyOrStaticPath])
+    : getLimitedPaths(
+        locationPaths.length === 0 ? [projectKeyOrStaticPath] : locationPaths
+      );
+  return {
+    suffix,
+    paths,
+    location,
```
Comment (CarlosCortizasCT) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:89]: Here we are returning the same parameter we received so I guess we can remove it since the function caller would already have it.

**Review by CarlosCortizasCT — COMMENTED** [review]
Great work over here! 💪 

I just left some comments with some suggestions.

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx (hunk 7 above)
Comment (ddouglasz) [reply packages/application-shell/src/components/application-page-title/application-page-title.tsx]: Thanks for all the feedback. 
updated here: https://github.com/commercetools/merchant-center-application-kit/pull/2758/commits/b422e3acf036da23e324d3dd1c701958ac9c7270

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+import { useLayoutEffect } from 'react';
+import { useLocation } from 'react-router-dom';
+import startCase from 'lodash/startCase';
+import capitalize from 'lodash/capitalize';
+
+type Breadcrumb<Query extends {} = {}> = {
+  suffix: string;
+  paths: string[];
+  location?: TEnhancedLocation<Query>;
```
Comment (CarlosCortizasCT) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:11]: I'm sorry but I don't understand why the `Breadcrumb` type includes this optional attribute since it doesn't seem is used anymore.
Perhaps I'm missing something. Could you please clarify? 🙏

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+};
+type Props = {
+  /**
+   * Allow to render a custom page title.
+   * Overrides default page title with the format:
+   * <?resource_name(manually inputed)> - <?page_location> - <application_identifier> - <project_key> - Merchant Center
```
Comment (CarlosCortizasCT) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:17]: I find a little confusing this explanation.
Perhaps we could go for something like this:
```
Allows to extend the page title. 
When provided, the array items will be used as a prefix for the default title this component will generate.

It will follow a pattern like this:
<renderPageTitle.join(' - ')> - <?page_location> - <application_identifier> - <project_key> - Merchant Center
```

What do you think?

Comment (ddouglasz) [reply packages/application-shell/src/components/application-page-title/application-page-title.tsx:17]: Oh yes, this makes sense too. I just wanted to provide some clarity to how the titles would be formed from an additional content provided.

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx (hunk 8 above)
Comment (ddouglasz) [reply packages/application-shell/src/components/application-page-title/application-page-title.tsx:11]: Honestly, I think I missed it, it is not used anymore after removing it from the returned object 😅
Thank you

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx (hunk 9 above)
Comment (ddouglasz) [reply packages/application-shell/src/components/application-page-title/application-page-title.tsx:16]: updated here: https://github.com/commercetools/merchant-center-application-kit/pull/2758/commits/07686cec816448de631b4144c58f8c5cd3eddc28

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+  suffix: string;
+  paths: string[];
+};
+type Props = {
+  /**
+   * Allow to render a custom page title.
+   * Overrides default page title with the format:
+   * <?resource_name(manually inputed)> - <?page_location> - <application_identifier> - <project_key> - Merchant Center
+   */
+  renderPageTitle?: string[];
```
Comment (emmenko) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:18]: Nit: the name of this prop seems a bit confusing to me. In the RFC I proposed `content` but we can also find another name as long as it fits with what it does.

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        ...paths.slice(2),
+      ],
+      true
+    );
+  }
+  return compilePath(paths, hasTruncatedPaths);
+};
+
+const isStaticPath = (path: string) => {
+  const staticPaths = ['account', 'login'];
```
Comment (emmenko) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:58]: Nit: Define this outside of the function, as it's a static variable.

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+      return title;
+    })
+    .join(' - ');
+};
+
+const ApplicationPageTitle = <Query extends {} = {}>(props: Props) => {
+  const location = useLocation();
+
+  useLayoutEffect(() => {
+    const breadcrumb = compilePageTitle(location as TEnhancedLocation<Query>);
```
Comment (emmenko) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:110]: I don't think we need `location` to be `TEnhancedLocation`, as we are only using the `pathname`. We also don't need the `Query` generic type.

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    const defaultMapping = `${startCase(breadcrumb.paths[0])} - ${
+      breadcrumb.suffix
+    }`;
+    const pageTitle =
+      props.renderPageTitle && props.renderPageTitle.length > 0
+        ? `${props.renderPageTitle
+            .join(' - ')
+            .replace(/\w+/g, capitalize)} - ${defaultMapping}`
+        : defaultMapping;
+    document.title = truncatePageTitle(pageTitle);
```
Comment (emmenko) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:120]: This logic seems a bit confusing, also because we're constructing `pageTitle` to then deconstruct it again in `truncatePageTitle`.

I would suggest to work with an array of parts, map over it and then join the values once.

One other thing I noticed: we shouldn't capitalize every word, especially the ones that have been passed as prop. I think only the `entryPointUriPath` could be capitalized, but not the rest.

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+ *
+ * It will follow a pattern like this:
+ * <renderPageTitle.join(' - ')> - <?page_location> - <application_identifier> - <project_key> - Merchant Center
+ *
+ */
+
+const compilePageTitle = <Query extends {}>(
+  location: TEnhancedLocation<Query>
+): Breadcrumb => {
+  const [, projectKeyOrStaticPath, ...locationPaths] =
```
Comment (emmenko) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:74]: Since we only look for the project key and the entry point, we only need to pick the first two:

```suggestion
  const [, projectKeyOrStaticPath, entryPointUriPath] =
```

image

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    const pageTitle =
+      props.renderPageTitle && props.renderPageTitle.length > 0
+        ? `${props.renderPageTitle
+            .join(' - ')
+            .replace(/\w+/g, capitalize)} - ${defaultMapping}`
+        : defaultMapping;
+    document.title = truncatePageTitle(pageTitle);
+  }, [location, props.renderPageTitle]);
+
+  return <></>;
```
Comment (emmenko) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:123]:
```suggestion
  return null;
```

**Review by emmenko — COMMENTED** [review]
Cool stuff, we're getting there 🙌

I did a quick review and left some feedback to some of the main points to focus on. Let me know if it's clear enough.

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    paths,
+  };
+};
+
+const truncatePageTitle = (pageTitle: string) => {
+  return pageTitle
+    .split(' - ')
+    .map((title) => {
+      if (title.length > maxTitleCharLength) {
+        return (title = `${title.slice(0, 12)} ... ${title.slice(
```
Comment (kark) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:97]: nit: I guess according to our rfc it should be 24 + 3 chars,  without empty characters
`This is a very long title that exceeds 24 characters → This is a ve...s characters`
```suggestion
        return (title = `${title.slice(0, 12)}...${title.slice(
```

Comment (kark) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:97]: nit: I guess according to our rfc it should be 24 + 3 chars,  without empty characters
`This is a very long title that exceeds 24 characters → This is a ve...s characters`

and we could also reference the `maxTitleCharLength` constant

```suggestion
        return (title = `${title.slice(0, maxTitleCharLength/2)}...${title.slice(
```

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx (hunk 15 above)
Comment (ddouglasz) [reply packages/application-shell/src/components/application-page-title/application-page-title.tsx]: Thank you, all feedback updated here: https://github.com/commercetools/merchant-center-application-kit/pull/2758/commits/e57096278223ab50a3baaffcb43433b5902c34b2

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx (hunk 16 above)
Comment (ddouglasz) [reply packages/application-shell/src/components/application-page-title/application-page-title.tsx]: Thank you Kacper 🙏:
Updated here: https://github.com/commercetools/merchant-center-application-kit/pull/2758/commits/e57096278223ab50a3baaffcb43433b5902c34b2

File: packages/application-shell/src/components/application-page-title/application-page-title.spec.tsx [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+import { render } from '@testing-library/react';
+import { Router } from 'react-router';
+import { createMemoryHistory } from 'history';
+
+const memoryHistory = (route: string) =>
+  createMemoryHistory({ initialEntries: [route] });
+
+describe.each`
+  route                                                                         | title
+  ${`/almond-40/dashboard`}                                                     | ${`Dashboard - almond-40 - Merchant Center`}
```
Comment (emmenko) [inline packages/application-shell/src/components/application-page-title/application-page-title.spec.tsx:12]: ⚠️ Let's use a fake project key

File: packages/application-shell/src/components/application-page-title/application-page-title.spec.tsx [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+          />
+        </Router>
+      );
+      await waitFor(() => {
+        expect(document.title).toBe(
+          'Some product...product name - Products - almond-40 - Merchant Center'
+        );
+      });
+    });
+  });
```
Comment (emmenko) [inline packages/application-shell/src/components/application-page-title/application-page-title.spec.tsx:84]: Nit: We can also use the `describe.each` here and have the test data in the "table" format.

File: packages/application-shell/src/components/application-shell-provider/application-shell-provider.tsx [hunk 19]
```diff
@@ -59,6 +60,13 @@ const ApplicationShellProvider = <AdditionalEnvironmentProperties extends {}>(
               <Suspense fallback={<ApplicationLoader />}>
                 <Router history={ApplicationShellProvider.history}>
                   <GtmBooter trackingEventList={props.trackingEventList || {}}>
+                    <ApplicationPageTitle
+                      content={[
+                        'A very long sentence that can be truncated',
+                        'Some red short text also kind of long',
+                      ]}
```
Comment (emmenko) [inline packages/application-shell/src/components/application-shell-provider/application-shell-provider.tsx:67]: Remember to remove this.

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+            }
+            return title;
+          })
+          .join(' - ');
+      } else {
+        return `${startCase(breadcrumb.paths[0])} - ${breadcrumb.suffix}`;
+      }
+    };
+
+    document.title = truncatedPageTitle();
```
Comment (emmenko) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:127]: I feel like we are overcomplicating things here.

I would expect something simple like:

```ts
const [, projectKeyOrStaticPath, entryPointUriPath] =
    location.pathname.split('/');

const customTitleParts = props.content.map((titlePart) => {
  if (titlePart.length <= 24) {
    return titlePart;
  }
  return [
    titlePart.slice(0, 12),
    titlePart.slice(titlePart.length - 12)
  ].join('...');
});

if (staticPaths.includes(projectKeyOrStaticPath)) {
  return [
    ...customTitleParts,
    upperFirst(projectKeyOrStaticPath),
    'Merchant Center'
  ].join('-');
}

return [
  ...customTitleParts,
  upperFirst(entryPointUriPath),
  projectKeyOrStaticPath,
  'Merchant Center'
].join('-');
```

**Review by emmenko — COMMENTED** [review]
I left some suggestions to iterate on the implementation, as it's a bit overcomplicated at the moment.

Thanks

**Comment by emmenko** [comment]
@ddouglasz I pushed a commit (https://github.com/commercetools/merchant-center-application-kit/pull/2758/commits/25406cf4ddcded3781d5cdf0d6a74de81a77b4f8) with some final improvements, mostly around the docs.

I also renamed the prop to `additionalParts`, as I felt `content` was not clear enough (sorry for suggesting that in the first place). I think this should make it more clear what it does now.

**Review by emmenko — APPROVED** [review]
This is good to go now from my side!

Thanks again for working on this 🤗

File: packages/application-shell/src/components/application-page-title/application-page-title.tsx [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+   * <ApplicationPageTitle additionalParts={['T-Shirt red']} />
+   * // T-Shirt red - Products - my-shop - Merchant Center
+   */
+  additionalParts: string[];
+};
+
+const maxTitleCharLength = 24;
+const staticPaths = ['account', 'login'];
+
+const defaultProps: Pick<TApplicationPageTitleProps, 'additionalParts'> = {
```
Comment (CarlosCortizasCT) [inline packages/application-shell/src/components/application-page-title/application-page-title.tsx:28]: nit: perhaps we don't need to use `Pick` helper anymore since the type only has one prop.

**Review by CarlosCortizasCT — APPROVED** [review]
Nice job 💯

**Review by kark — APPROVED** [review]
